cwlVersion: v1.2
class: Workflow
label: Transgene Integration LR workflow v1.4
$namespaces:
  sbg: https://sevenbridges.com

requirements:
- class: SubworkflowFeatureRequirement
- class: MultipleInputFeatureRequirement
- class: StepInputExpressionRequirement
- class: InlineJavascriptRequirement

inputs:
  - id: fastqfile
    type: File
    label: Input FASTQ File
    doc: Input FASTQ/FASTQ.GZ file for the sample. Currently only 1 FASTQ file is allowed per sample. Merge files into one before running this pipeline if there are multiple FASTQ files for one sample.
  - id: reffile
    type: File
    label: Reference Sequence File
    doc: Reference sequence file should contain possible sequences for WT (Unmodified), HDR, ITR integration. 
  - id: sample_name
    type: string
    label: Sample Name
    doc: Sample Name is used for naming output files.
  - id: primer_1
    type: string?
    label: Forward Primer Sequence
    doc: Forward primer sequence in 5'->3' direction.
  - id: primer_2
    label: Reverse Primer Sequence
    doc: Reverse primer sequence in 5'->3' direction.
    type: string?
  - id: config
    type: File
    label: Transgene Config File
    doc: Config file listing detailed information for all reference sequences provided in the Reference file. Should include 9 columns, including AAV_Vector, Ref_Name, Ref_Type, ITR1_start, ITR1_end, gene_start, gene_end, ITR2_start, ITR2_end. A header line is required for the config file.
  - id: truncated_cutoff
    type: float?
    label: Truncated HDR/ITR Threshold
    doc: Truncated HDR/ITR threshold. Should be 0-1, default is 0.99.
  - id: min_qual
    type: float?
    label: Minimal average base quality
    doc: Sequencing reads with average Phred base quality below this cutoff will be filtered out. Default is 30. 
  - id: five_prime_HA_seq
    type: string
    label: 5' Homology Arm Sequence
    doc: 5' Homology Arm sequence in 5'->3' direction.
  - id: three_prime_HA_seq
    type: string
    label: 3' Homology Arm Sequence
    doc: 3' Homology Arm sequence in 5'->3' direction.
  - id: HA_Seq_match_ratio
    type: float?
    default: 0.98
    label: Homology Arm Sequence Match Ratio
    doc: Homology Arm Sequence Match Ratio. Default is 0.98.
  - id: left_itr_seq
    type: string?
    label: Left Side ITR Sequence
    doc: Left ITR sequence in 5'->3' direction.
  - id: right_itr_seq
    type: string?
    label: Right Side ITR Sequence
    doc: Right ITR sequence in 5'->3' direction.
  - id: seed_size
    type: int?
    default: 15
    label: Seed Size
    doc: Seed size for blastn. Default is 15.
  - id: perc_identity
    type: int?
    default: 80
    label: Percent Identity
    doc: Min percent identity to match a ITR sequence for blastn. Default is 80.

outputs:
  - id: grouped_outputs
    type: Directory
    label: grouped output folder
    doc: A folder that contains all output files generated by the workflow 
    outputSource: group_output/outputs
  - id: count_table
    type: File
    outputSource: count_read/count_table

steps:
  - id: filter_primer
    label: Filter reads
    in:
    - id: fastqfile
      source: fastqfile
    - id: primer_f
      source: primer_1
    - id: primer_r
      source: primer_2
    - id: sample_name
      source: sample_name
    - id: min_qual
      source: min_qual
    run: aav_per_sample_workflow.cwl.steps/filter_primer.cwl
    out:
    - id: filtered_fastq
    - id: filter_log
    - id: original_histogram_plot
    - id: filtered_histogram_plot

  - id: minimap2
    label: Minimap2 Read Alignment
    in:
    - id: fastqfile
      source: filter_primer/filtered_fastq
    - id: reffile
      source: reffile
    run: aav_per_sample_workflow.cwl.steps/align_minimap2.cwl
    out:
    - id: aligned_sam

  - id: samtools_sort
    label: Samtools Sort
    in:
    - id: samfile
      source: minimap2/aligned_sam
    - id: sample_name
      source: sample_name
    run: aav_per_sample_workflow.cwl.steps/samtools_sort.cwl
    out:
    - id: sorted_bam

  - id: classification
    label: Read Classification
    in:
    - id: bamfile
      source: samtools_sort/sorted_bam
    - id: reffile
      source: reffile
    - id: sample_name
      source: sample_name
    - id: config
      source: config
    - id: truncated_cutoff
      source: truncated_cutoff
    - id: filtered_fastq
      source: filter_primer/filtered_fastq
    - id: five_prime_HA_seq
      source: five_prime_HA_seq
    - id: three_prime_HA_seq
      source: three_prime_HA_seq
    - id: HA_Seq_match_ratio
      source: HA_Seq_match_ratio
    - id: left_itr_seq
      source: left_itr_seq
    - id: right_itr_seq
      source: right_itr_seq
    - id: seed_size
      source: seed_size
    - id: perc_identity
      source: perc_identity
    run: aav_per_sample_workflow.cwl.steps/classify_read.cwl
    out:
    - id: read_classification

  - id: count_read
    label: Count Read
    in:
    - id: output_name
      source: sample_name
    - id: classification
      source: classification/read_classification
    run: aav_per_sample_workflow.cwl.steps/count_read.cwl
    out:
    - id: count_table
    - id: pie_chart

  - id: group_output
    label: Group output files to a folder
    in:
      - id: output_name
        source: sample_name
      - id: in_array
        linkMerge: merge_flattened
        source:
          - filter_primer/filter_log
          - filter_primer/filtered_fastq
          - filter_primer/original_histogram_plot
          - filter_primer/filtered_histogram_plot
          - samtools_sort/sorted_bam
          - classification/read_classification
          - count_read/count_table
          - count_read/pie_chart
    run: aav_per_sample_workflow.cwl.steps/group_output.cwl
    out:
      - id: outputs
